### DISCUSSION TITLE
How can i pass 2FA authentification?

### DISCUSSION THREAD

**Discussion by Soulbadguy54** [ValvePython/steam#325] [Q&A] [answered]
Hello. For now every run of the script requires 2FA steam guard code. It makes the use of the script impossible. Because obviously i want to use it in backgroud with the possibility of autorestarts.
How can i pass 2FA steam guard? Probably make some guard file? But where can i get it? What should be in it? Where should i pass it in this library? 
Thx for answers! <3 

Some additionals, I already get secret_share and identify_secret. How can i use it in this lib to pass 2FA ?

**Comment by rossengeorgiev** [comment, marked as answer]
There are two options:

1. Use `SteamAuthenticator` and generate a 2FA for every login

```python
from steam.client import SteamClient
from steam.guard import SteamAuthenticator

client = SteamClient()
client.set_credential_location('./secrets_directory')

code = SteamAuthenticator(secrets={"shared_secret": "...."}).get_code()

client.login(username, password, two_factor_code=code)
```

2. Use `login_key` (Steam remember password feature)

```python
# first login
client1 = SteamClient()
client1.set_credential_location('.')
client1.cli_login()

if not client1.relogin_available:
    client1.wait_event(EMsg.ClientNewLoginKey)

# save credentials for future logins
relogin_credentials = client1.username, client1.login_key

# second login (for example, reconnecting after being disconnected)
client1.disconnect()
client1.relogin()  # uses username and login_key

# third login (login after a while)

client2 = SteamClient()
client2.set_credential_location('.')
client2.username, client2.login_key = relogin_credentials
client2.relogin()
```

Example from `steamctl` handling remembered credentials:

https://github.com/ValvePython/steamctl/blob/63ab5f548eeefbf087c9feed5d33ff5e2d52ca6b/steamctl/clients.py#L65-L85

**Comment by Soulbadguy54** [comment]
Thx a lot! The first way looks slightly better. Do i really need the row with credentials? Looks like steam auth code is all i need :)

**Reply by rossengeorgiev** [reply]
Up to you

**Comment by Kein** [comment]
Am I missing something?

```python
client = SteamClient()
userdata = Path(USER_FILE)
if userdata.is_file():
    # read the data, populate SteamClient instance with user and login_key
    result = client.relogin()
    print(result)
else:
    client.cli_login()
```

Always returns LOGONDENIED.

**Reply by rossengeorgiev** [reply]
I assume you get the result on your `relogin()` call, and userdata is previously saved `login_key`. If so, then first set the value back on the client instance before calling `relogin()`. To use `relogin()` you need `username` and `login_key` set from previous successful login

**Reply by Kein** [reply]
That's what I do? If saved file with `username` and `login_key` exist I read the file, populate `username` and `login_key` on SteamClient instance and then call `relogin()`

It always results in `LOGONDENIED`.
The `login_key` can't be outdated because it was received in previous session and saved between script re-runs.

**Reply by rossengeorgiev** [reply]
Are you saying the example from https://github.com/ValvePython/steam/discussions/325#discussioncomment-611865 is not working for you? I've just updated it to make it more clear

**Reply by Kein** [reply]
This part:
```python
client2 = SteamClient()
client2.username, client2.login_key = relogin_credentials
client2.relogin()
```
Does not. I run script on cron each x minutes, at the start I read saved login_key from previous login, populate client instance and then call to relogin() and it always results in `LOGONDENIED`

**Reply by rossengeorgiev** [reply]
And if you add `client1.credential_location('.')` and `client2.credential_location('.')` ?

**Reply by Kein** [reply]
Not sure I understand? When I run script a new I only have one client instance I've just created. The only thing from past run I have is saved to storage login_key and username.

**Reply by rossengeorgiev** [reply]
If you are on email 2fa, or no 2fa, you may need the sentry blob. `credential_location` set the location where that file will be saved/loaded from. Just set the lines above just after `clientx = SteamClient()` and rerun.

**Reply by Kein** [reply]
Can you elaborate, you got me a bit confused here.

**Reply by rossengeorgiev** [reply]
updated the top example

**Reply by Kein** [reply]
There seems to be error with your example as `.credential_location` is a property, not a function, but I figured it out and yes, it now saves sentry file and works, thanks.

**Reply by rossengeorgiev** [reply]
It's `set_credential_location()`. I've fixed the example

**Comment by Kein** [comment]
How do I properly do re-login when steam rejects my sentry data?
```py
client = SteamClient()
client.credential_location = '.'
userdata = Path(USER_FILE)
result = None
if userdata.is_file():
    f = open(USER_FILE, 'r', encoding='utf-8')
    client.username = f.readline().strip().rstrip('\n')
    client.login_key = f.readline().strip().rstrip('\n')
    f.close()
    result = client.relogin()
    #print(result)
else:
    client.cli_login()

if not client.relogin_available:
    client.wait_event(EMsg.ClientNewLoginKey)
```
the example says I need to wait for `EMsg.ClientNewLoginKey` but it throws with "This operation will block forever"
In fact, I dont even understand why does sentry data expire in like few minutes after being created.
